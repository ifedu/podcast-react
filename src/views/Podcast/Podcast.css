.Podcast {
  display: flex;
  justify-content: space-between;
}

.Podcast .info {
  box-shadow: 0 1px 2px 2px #ddd;
  width: 300px;
}

.Podcast .info > div {
  padding: 8px;
}

.Podcast .table {
  width: 800px;
}

.Podcast .episode > div {
  box-shadow: 0 1px 2px 2px #ddd;
}

.Podcast .episode > div:nth-child(1) {
  font-size: 20px;
  font-weight: bold;
  margin-bottom: 16px;
  padding: 12px;
}

.Podcast .episode .table {
  padding: 24px 16px;
}

.Podcast .episode .head {
  font-weight: bold;
}

.Podcast .episode .row {
  border-bottom: 1px solid #ddd;
  border-top: 1px solid #ddd;
  display: grid;
  grid-template-columns: 75% 15% 10%;
  padding: 0px 16px 16px 0;
}
