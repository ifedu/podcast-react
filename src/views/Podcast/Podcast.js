import React from 'react';
import { Link, useLocation, useParams } from 'react-router-dom';

import './Podcast.css';
import { useLoading } from '../../hooks/LoadingContext';
import { usePodcast } from '../../hooks/usePodcast';
import { PodcastInfo } from '../../components/PodcastInfo/PodcastInfo';

function convertToShortDate(fullDate) {
  const date = new Date(fullDate);
  const day = date.getDate();
  const month = date.getMonth();
  const year = date.getFullYear();

  return `${day}/${Number(month) + 1}/${year}`;
}

function convertToTime(ms) {
  const date = new Date(ms);
  const minutes = date.getMinutes();
  const seconds = date.getSeconds();

  return `${minutes}:${seconds}`;
}

export function Podcast() {
  const { setIsLoading } = useLoading();
  const { state: { song } } = useLocation();
  const { podcastId } = useParams();

  const podcastInfo = usePodcast(setIsLoading, podcastId);
  if (!podcastInfo) {
    return null;
  }

  return (
    <div className='Podcast' data-testid='podcast'>
      <PodcastInfo info={podcastInfo.results[0]} song={song}/>

      <div className='episode'>
        <div>Episodes: {podcastInfo.resultCount}</div>

        <div className='table'>
          <div className='head row'>
            <span>Title</span>
            <span>Date</span>
            <span>Duration</span>
          </div>

          <div className='body'>
            {
              podcastInfo.results.map((episode, i) => {
                if (!episode.description) return null;

                return <Link key={i} to={`/podcast/${podcastId}/episode/${episode.trackId}`} state={{ podcastInfo, song, episode }} data-testid='link-to-episode'>
                  <div className='episode row'>
                    <span>{episode.trackName}</span>
                    <span>{convertToShortDate(episode.releaseDate)}</span>
                    <span>{convertToTime(episode.trackTimeMillis)}</span>
                  </div>
                </Link>
              })
            }
          </div>
        </div>
      </div>
    </div>
  );
}
