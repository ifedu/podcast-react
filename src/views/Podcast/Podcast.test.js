import { cleanup, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes, useParams } from 'react-router-dom';
import { create } from 'react-test-renderer' ;

import { Podcast } from './Podcast';

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

jest.mock('../../hooks/LoadingContext.js', () => ({
  useLoading: () => ({
    setIsLoading: jest.fn(),
  }),
}));

jest.mock('../../hooks/usePodcast.js', () => ({
  usePodcast: () => ({
    resultCount: 1,
    results: [
      {
        description: 'Podcast Description',
        trackId: 1,
      }
    ]
  }),
}));

const getContent = (pathname = '/') => {
  const state = {
    song: {
      summary: {},
    },
  };

  return <MemoryRouter initialEntries={[{ pathname, state }]}>
    <Routes>
      <Route path='/podcast/1' element={
        <Podcast/>
      }></Route>

      <Route path='/podcast/1/episode/1' element={
        <div data-testid='episode'></div>
      }></Route>
    </Routes>
  </MemoryRouter>;
}

describe('Podcast', () => {
  let view = getContent('/podcast/1');

  beforeEach(() => {
    useParams.mockReturnValue({ podcastId: 1 });
  });

  afterEach(cleanup);

  it('renders correctly', () => {
    const tree = create(view).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the podcast', async() => {
    render(view);

    await waitFor(() => {
      const podcast = screen.getByTestId('podcast');
      expect(podcast).toBeInTheDocument();
    });
  });

  it('navigates from podcast page to the episode page', async () => {
    render(view);

    await waitFor(() => {
      const podcast = screen.getByTestId('podcast');
      expect(podcast).toBeInTheDocument();
    });

    let link = screen.getByTestId('link-to-episode');
    userEvent.click(link);

    await waitFor(() => {
      const episode = screen.getByTestId('episode');
      expect(episode).toBeInTheDocument();
    });
  });
});
