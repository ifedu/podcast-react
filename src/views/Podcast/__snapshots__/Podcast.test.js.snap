// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcast renders correctly 1`] = `
<div
  className="Podcast"
  data-testid="podcast"
>
  <div
    className="PodcastInfo"
    data-testid="podcast-info"
  >
    <a
      href="/podcast/undefined"
      onClick={[Function]}
    >
      <div
        className="section picture"
      >
        <img
          alt="All Songs Considered"
        />
      </div>
      <div
        className="section title"
      >
        <span />
        <div>
          by 
        </div>
      </div>
    </a>
    <div
      className="section description"
    >
      <span>
        Description
      </span>
      <div />
    </div>
  </div>
  <div
    className="episode"
  >
    <div>
      Episodes: 
      1
    </div>
    <div
      className="table"
    >
      <div
        className="head row"
      >
        <span>
          Title
        </span>
        <span>
          Date
        </span>
        <span>
          Duration
        </span>
      </div>
      <div
        className="body"
      >
        <a
          data-testid="link-to-episode"
          href="/podcast/1/episode/1"
          onClick={[Function]}
        >
          <div
            className="episode row"
          >
            <span />
            <span>
              NaN/NaN/NaN
            </span>
            <span>
              NaN:NaN
            </span>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
